client/packages/core/src/utils/PersistedObject.ts

```diff
@@ -23,155 +27,521 @@ function safeIdleCallback(cb, timeout: number) {
+    }
+
+    const keysToLoad = [];
+
+    const kvPairs: Array<[string, any]> = [[META_KEY, metaValue]];
+    const metaObjects: Meta<K>['objects'] =
+      metaValue.objects ?? ({} as Meta<K>['objects']);
+    metaValue.objects = metaObjects;
+    for (const k of keysToUpdate) {
+      if (this._loadedKeys.has(k)) {
+        const serializedV = this.serialize(k, this.currentValue[k]);
+        kvPairs.push([k, serializedV]);
+        const size = this._objectSize(serializedV);
+        const m = metaObjects[k] ?? {
+          createdAt: Date.now(),
+          updatedAt: Date.now(),
+          size,
+        };
+        m.updatedAt = Date.now();
+        m.size = size;
+        metaObjects[k] = m;
+        this._pendingSaveKeys.delete(k);
+      } else {
+        keysToLoad.push(k);
+      }
+    }
+
+    const p = this._persister.multiSet(kvPairs);
+    promises.push(p.then(() => 1));
+
+    // For the keys that haven't loaded, load the key then try
+    // persisting again. We don't want to do any async work here
+    // or else we might end up saving older copies of the data to
+    // the store.
+    for (const k of keysToLoad) {
+      const p = this._loadKey(k).then(() => this._enqueuePersist(opts));
+      promises.push(p);
+    }
+
+    if (!skipGc) {
+      this.gc();
     }
-    this._pendingSaveCbs.length = 0;
+
+    return Promise.all(promises).then((vs) => {
+      return vs.reduce((acc, x) => acc + x, 0);
+    });
   }
 
   async flush() {
     if (!this._nextSave) {
       return;
     }
     clearTimeout(this._nextSave);
-    this._writeToStorage();
+    this._nextSave = null;
+    const p = this._writeToStorage();
+    return p;
   }
 
-  _enqueuePersist(cb) {
-    if (this._nextSave) {
-      if (cb) {
-        this._pendingSaveCbs.push(cb);
+  private async _gc() {
+    const keys = new Set(await this._persister.getAllKeys());
+    keys.delete(META_KEY);
+
+    // Keys we can't delete
+    const sacredKeys = new Set(Object.keys(this.currentValue));
+    for (const k of Object.keys(this._loadingKeys)) {
+      sacredKeys.add(k);
+    }
+    for (const k of this._loadedKeys) {
+      sacredKeys.add(k);
+    }
+
+    // Refresh meta from the store so that we're less likely to
+    // clobber data from other tabs
+    const meta = await this._refreshMeta();
+    if (!meta) {
+      this._log.info('Could not gc because we were not able to load meta');
+      return;
+    }
+
+    const promises = [];
+
+    const deets = {
+      gcOpts: this._gcOpts,
+      keys,
+      sacredKeys,
+      removed: [],
+      metaRemoved: [],
+      removedMissingCount: 0,
+      removedOldCount: 0,
+      removedThresholdCount: 0,
+      removedSizeCount: 0,
+    };
+
+    // First, remove all keys we don't know about
+    for (const key of keys) {
+      if (sacredKeys.has(key) || key in meta.objects) {
+        continue;
       }
+      this._log.info('Lost track of key in meta', key);
+      promises.push(this._persister.removeItem(key));
```

Comment:
One worry I have here: 

https://react-native-async-storage.github.io/2.0/API/#getallkeys

I _think_ async storage will not just return keys that we save, but return _all_ keys that the user saves in their async storage. 

Would this potentially delete their items?

Reply:
Update: just saw that you wrote a filter in the native js! Carry on sir 🫡🫡🫡